cpp/src/arrow/dataset/file_json.cc

```diff
@@ -0,0 +1,414 @@
+
+  std::vector<bool> selected(max_num_fields, false);
+  std::vector<int> toplevel_indices;
+  toplevel_indices.reserve(max_num_fields);
+
+  for (const auto& ref : scan_options.MaterializedFields()) {
+    auto index = TopLevelIndex(ref, *dataset_schema);
+    DCHECK_GE(index, 0);
+    if (selected[index]) continue;
+
+    // Determine if the field exists in the physical schema before selecting it
+    bool found;
+    if (!ref.IsNested()) {
+      const auto& name = dataset_schema->field(index)->name();
+      found = physical_schema.GetFieldIndex(name) != -1;
+    } else {
+      // Check if the nested field is present in the physical schema. If so, we load its
+      // entire top-level field
+      ARROW_ASSIGN_OR_RAISE(auto path, ref.FindOne(*dataset_schema));
+      auto universal_ref = ToUniversalRef(path, *dataset_schema);
+      ARROW_ASSIGN_OR_RAISE(auto match, universal_ref.FindOneOrNone(physical_schema));
+      found = !match.empty();
+    }
+
+    if (!found) continue;
+
+    toplevel_indices.push_back(index);
+    selected[index] = true;
+    // All fields in the dataset schema require materialization, so return early
+    if (toplevel_indices.size() == max_num_fields) {
+      return dataset_schema;
+    }
+  }
+
+  FieldVector fields;
+  fields.reserve(toplevel_indices.size());
+  std::sort(toplevel_indices.begin(), toplevel_indices.end());
+  for (auto index : toplevel_indices) {
+    fields.push_back(dataset_schema->field(index));
+  }
+
+  return ::arrow::schema(std::move(fields));
+}
+
+Result<std::shared_ptr<JsonFragmentScanOptions>> GetJsonFormatOptions(
+    const JsonFileFormat& format, const ScanOptions* scan_options) {
+  return GetFragmentScanOptions<JsonFragmentScanOptions>(
+      kJsonTypeName, scan_options, format.default_fragment_scan_options);
+}
+
+Result<Future<ReaderPtr>> DoOpenReader(
+    const FileSource& source, const JsonFileFormat& format,
+    const std::shared_ptr<ScanOptions>& scan_options = nullptr) {
+  ARROW_ASSIGN_OR_RAISE(auto json_options,
+                        GetJsonFormatOptions(format, scan_options.get()));
+
+  struct State {
+    State(const JsonFragmentScanOptions& json_options,
+          const std::shared_ptr<ScanOptions>& scan_options)
+        : parse_options(json_options.parse_options),
+          read_options(json_options.read_options),
+          scan_options(scan_options) {
+      // We selectively ignore some user options, primarily those that will influence the
+      // output schema.
+      parse_options.explicit_schema = nullptr;
+      parse_options.unexpected_field_behavior = json::UnexpectedFieldBehavior::InferType;
+      // We don't wish to contend with the scanner's CPU threads if multiple fragments are
+      // being scanned in parallel, so fragment-level parallelism gets conditionally
+      // disabled.
+      if (this->scan_options && this->scan_options->use_threads) {
+        read_options.use_threads = false;
+      }
```

Comment:
We've done this in the past to avoid nested deadlock.  That shouldn't be a problem if we use the async paths (though I don't know what the async path for the JSON reader looks like).

If there are only 1 or 2 files and there is a potential to speed up JSON decoding with multiple threads then that seems like something we'd want to do.

Reply:
There was a similar comment in `file_csv.cc` that I based that on (since the CSV reader has similar async properties). Admittedly, I wasn't completely on board with the reasoning myself.

Otherwise, I don't think there's any real risk of nested parallelism.